### ISSUE TITLE
CommandTimeout on Ef core migration

### ISSUE THREAD

**Issue by incolor2** [dotnet/efcore#22887]
Hi,
I have a long running db migration that times out. I tried to raise the command timeout but I am unable to have it take effect. No matter what i set the ef migration is logging CommandTimeout = 4000.
So i guess my question is how do i raise this timeout?

The code
```C#
            services.AddDbContext<PlanSearchContext>(options =>
                options.UseMySql(connectionString, mysqlOptions =>
                {
                    //Setting DB version
                    mysqlOptions.ServerVersion(new Version(5, 7), ServerType.MySql);

                    //Some migrations are taking a looong time.
                    mysqlOptions.CommandTimeout(18000);                    
                }

                ));
```

The dotnet ef migrations log

```
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.8 initialized 'PlanSearchContext' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: CommandTimeout=18000 MigrationsAssembly=Scania.Maintenance.PlanSearch.DbMigration ServerVersion 5.7 MySql
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (67ms) [Parameters=[], CommandType='Text', CommandTimeout='4000']
      SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='PlanSearchDb' AND TABLE_NAME='__EFMigrationsHistory';
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (47ms) [Parameters=[], CommandType='Text', CommandTimeout='4000']
      SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='PlanSearchDb' AND TABLE_NAME='__EFMigrationsHistory';
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (50ms) [Parameters=[], CommandType='Text', CommandTimeout='4000']
      SELECT `MigrationId`, `ProductVersion`
      FROM `__EFMigrationsHistory`
      ORDER BY `MigrationId`;
info: Microsoft.EntityFrameworkCore.Migrations[20402]
      Applying migration '20200924064330_SMA1097'.
Applying migration '20200924064330_SMA1097'.
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (1,365,543ms) [Parameters=[], CommandType='Text', CommandTimeout='4000']
      ALTER TABLE `ModuleTypes` ADD `CommonVersionId` binary(16) NOT NULL DEFAULT 0x00000000000000000000000000000000;
fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
      Failed executing DbCommand (4,000,062ms) [Parameters=[], CommandType='Text', CommandTimeout='4000']
      ALTER TABLE `Modules` ADD `CommonVersionId` binary(16) NOT NULL DEFAULT 0x00000000000000000000000000000000;
Failed executing DbCommand (4,000,062ms) [Parameters=[], CommandType='Text', CommandTimeout='4000']
ALTER TABLE `Modules` ADD `CommonVersionId` binary(16) NOT NULL DEFAULT 0x00000000000000000000000000000000;
MySql.Data.MySqlClient.MySqlException (0x80004005): The Command Timeout expired before the operation completed.
 ---> MySql.Data.MySqlClient.MySqlException (0x80004005): The Command Timeout expired before the operation completed.
 ---> System.IO.IOException: Unable to read data from the transport connection: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond..
 ---> System.Net.Sockets.SocketException (10060): A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.Security.SslStream.FillBufferAsync[TReadAdapter](TReadAdapter adapter, Int32 minSize)
   at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
   at System.Net.Security.SslStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.Stream.Read(Span`1 buffer)
   at MySqlConnector.Utilities.Utility.Read(Stream stream, Memory`1 buffer) in /_/src/MySqlConnector/Utilities/Utility.cs:line 439
   at MySqlConnector.Protocol.Serialization.StreamByteHandler.<ReadBytesAsync>g__DoReadBytesSync|6_0(Memory`1 buffer_) in /_/src/MySqlConnector/Protocol/Serialization/StreamByteHandler.cs:line 39
   at MySqlConnector.Protocol.Serialization.BufferedByteReader.ReadBytesAsync(IByteHandler byteHandler, ArraySegment`1 buffer, Int32 totalBytesToRead, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/BufferedByteReader.cs:line 36
   at MySqlConnector.Protocol.Serialization.ProtocolUtility.ReadPacketAsync(BufferedByteReader bufferedByteReader, IByteHandler byteHandler, Func`1 getNextSequenceNumber, ProtocolErrorBehavior protocolErrorBehavior, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/ProtocolUtility.cs:line 409
   at MySqlConnector.Protocol.Serialization.ProtocolUtility.DoReadPayloadAsync(BufferedByteReader bufferedByteReader, IByteHandler byteHandler, Func`1 getNextSequenceNumber, ArraySegmentHolder`1 previousPayloads, ProtocolErrorBehavior protocolErrorBehavior, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/ProtocolUtility.cs:line 467
   at MySqlConnector.Protocol.Serialization.ProtocolUtility.ReadPayloadAsync(BufferedByteReader bufferedByteReader, IByteHandler byteHandler, Func`1 getNextSequenceNumber, ArraySegmentHolder`1 cache, ProtocolErrorBehavior protocolErrorBehavior, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/ProtocolUtility.cs:line 462
   at MySqlConnector.Protocol.Serialization.StandardPayloadHandler.ReadPayloadAsync(ArraySegmentHolder`1 cache, ProtocolErrorBehavior protocolErrorBehavior, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/StandardPayloadHandler.cs:line 42
   at MySqlConnector.Core.ServerSession.ReceiveReplyAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 766
--- End of stack trace from previous location where exception was thrown ---
   at MySqlConnector.Core.ServerSession.ReceiveReplyAsyncAwaited(ValueTask`1 task) in /_/src/MySqlConnector/Core/ServerSession.cs:line 794
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 51
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet() in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlDataReader.cs:line 130
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlDataReader.cs:line 391
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 62
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlCommand.cs:line 218
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery() in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlCommand.cs:line 68
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Pomelo.EntityFrameworkCore.MySql.Migrations.Internal.MySqlMigrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
The Command Timeout expired before the operation completed.
```

EF Core version: Ef 3.1.8, Entity Framework Core .NET Command-line Tools 3.1.8, 
Database provider: Pomelo.EntityFrameworkCore.MySql
Target framework: dotnet core 3.1
Operating system: amazon linux
IDE: Visual Studio 2019 16.4.5

**Comment by ajcvickers** [comment]
@lauxjpn We're not sure if this is a Pomelo issue or an EF Core issue--any thoughts? Also, CommandTimeout is documented to take seconds; these are curiously large values... Does MySQL do something different here?

**Comment by lauxjpn** [comment]
> Also, CommandTimeout is documented to take seconds; these are curiously large values... Does MySQL do something different here?

@ajcvickers No, Pomelo uses seconds as well (we just forward the call to the `RelationalOptionsExtension.WithCommandTimeout()` implementation).

---

@incolor2 A couple of questions to further investigate:
* Is this a private or a public project?
* Is the failing migration being applied to a production or a development database?
* Is the database local or remote (from the point of view of where the `dotnet ef` command is being run)?
* How does the `CREATE TABLE` script of the `Modules` table look like (generated by MySQL, `SHOW CREATE TABLE`) before applying the migration?
* How many rows are in the `Modules` table?
* Have you tried to execute the failing `ALTER TABLE` statement manually in the MySQL console or MySQL Workbench or similar tools? If so, how was the outcome?
* Have you ensured, that the table is not corrupt (`CHECK TABLE` and `REPAIR TABLE`)?
* Are there any other processes accessing the table (could be from other machines) at the same time you try to apply the migration?

**Comment by incolor2** [comment]
Hi guys,

Is this a private or a public project?
- It is unfortunately a private one

Is the failing migration being applied to a production or a development database?
- The dev one is running smooth (Very little data). But when I run it in a prod like env it fails.

Is the database local or remote (from the point of view of where the dotnet ef command is being run)?
- The DB is remote. It is running in that "other" cloud (Amazon Aurora). 
The ef command is executed from the build/deploy pipeline. But the same result if I run it from my dev machine.

How does the CREATE TABLE script of the Modules table look like (generated by MySQL, SHOW CREATE TABLE) before applying the migration?
- I will check and get back with this.

How many rows are in the Modules table?
- 200 Million ish..

Have you tried to execute the failing ALTER TABLE statement manually in the MySQL console or MySQL Workbench or similar tools? If so, how was the outcome?
- Yes i did run it in the MySQL Workbench but to be honest i did hit the "MySQL Workbench" timeout after 2 hrs (not 4000s). I will raise the timeout and run again.

Have you ensured, that the table is not corrupt (CHECK TABLE and REPAIR TABLE)?
- Yes the table looks ok

Are there any other processes accessing the table (could be from other machines) at the same time you try to apply the migration?
- Yes it could be. We did not close the application


It is a big table and it seems to take a very long time to run this operation. The column is added so i am guessing the command is running until competition on the server. But the things it should do after will fail.

EF is doing exakt what it says it is going to do. It is timing out after 4000s. But where does that value come from? It does not seem to be a random timeout.

**Comment by lauxjpn** [comment]
> It is a big table and it seems to take a very long time to run this operation. The column is added so i am guessing the command is running until competition on the server. But the things it should do after will fail.

Thanks for answering. A bit of context is usually helpful.

> EF is doing exakt what it says it is going to do. It is timing out after 4000s. But where does that value come from? It does not seem to be a random timeout.

I'll take a look at this. I will also do some tests regarding performance of the ALTER TABLE operation regarding MySQL and Aurora.

---

You might want to take a look at [Altering Tables in Amazon Aurora Using Fast DDL](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Managing.FastDDL.html), to increase performance:

> In Amazon Aurora, you can use fast DDL to run an `ALTER TABLE` operation in place, nearly instantaneously.

**Comment by incolor2** [comment]
Here is the table

CREATE TABLE `Modules` (
  `Id` binary(16) NOT NULL,
  `ModuleTypeId` binary(16) NOT NULL,
  `PerformedDate` datetime(6) DEFAULT NULL,
  `PerformedDistance` int(11) DEFAULT NULL,
  `Status` int(11) NOT NULL,
  `PlannedEndDate` datetime(6) DEFAULT NULL,
  `WorkshopScheduleDate` datetime(6) DEFAULT NULL,
  `CommonVersionId` binary(16) NOT NULL DEFAULT '\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0',
  PRIMARY KEY (`Id`),
  KEY `IX_Modules_Id` (`Id`),
  KEY `IX_Modules_ModuleTypeId` (`ModuleTypeId`),
  CONSTRAINT `FK_Modules_ModuleTypes_ModuleTypeId` FOREIGN KEY (`ModuleTypeId`) REFERENCES `ModuleTypes` (`Id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=latin1;

**Comment by lauxjpn** [comment]
@incolor2

**Comment by lauxjpn** [comment]
> What is the exact version of ~~MySQL and~~ Pomelo when applying the migration?

@incolor2 Please also post the version of Pomelo that is being used when running the `dotnet ef` command.

**Comment by lauxjpn** [comment]
I did some tests about the `CommandTimeout` option with Pomelo 3.2.2. It is being honored (at least it is shown in the command line) as it shouldbe.

Here I used `18000` seconds (5 hours):

```c#
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
{
    optionsBuilder
        .UseMySql("server=127.0.0.1;port=3307;user=root;password=;database=EFCoreIssue22887",
            b => b.ServerVersion("5.7.0-mysql")
                  .CommandTimeout(18000)) // 5 hours
        .UseLoggerFactory(LoggerFactory.Create(b => b
            .AddConsole()
            .AddFilter(level => level >= LogLevel.Information)))
        .EnableSensitiveDataLogging()
        .EnableDetailedErrors();
}
```
The logging output looks like this:
```
PS E:\Sources\PomeloIssues\EFCoreIssue22887\IssueConsoleTemplate> dotnet ef database update                                                                                                                                                                                    [...]
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.8 initialized 'Context' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: CommandTimeout=18000 ServerVersion 5.7.0 MySql SensitiveDataLoggingEnabled DetailedErrorsEnabled
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (3ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      CREATE DATABASE `EFCoreIssue22887`;
[...]
```

**Comment by lauxjpn** [comment]
I also did some performance testing against MySQL 5.7.31 with the following code:

<details>
<summary>Sample program used for simple performance test</summary>

```c#
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
using Microsoft.EntityFrameworkCore.ValueGeneration;

// 1. Add initial migration.
// 2. Update database.
// 3. Add test data.
// 4. Add new migration.
// 5. Update database.

namespace IssueConsoleTemplate
{
    public class Module
    {
        public Guid Id { get; set; }
        public Guid ModuleTypeId { get; set; }
        public DateTime? PerformedDate { get; set; }
        public int? PerformedDistance { get; set; }
        public int Status { get; set; }
        public DateTime? PlannedEndDate { get; set; }
        public DateTime? WorkshopScheduleDate { get; set; }
        
        // 4. Add in new migration:
        // public Guid CommonVersionId { get; set; }
        
        public ModuleType ModuleType { get; set; }
    }
    
    public class ModuleType
    {
        public Guid Id { get; set; }

        // 4. Add in new migration:
        // public Guid CommonVersionId { get; set; }

        public virtual ICollection<Module> Modules { get; set; } = new HashSet<Module>();
    }

    public class Context : DbContext
    {
        public virtual DbSet<Module> Modules { get; set; }
        public virtual DbSet<ModuleType> ModuleTypes { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                .UseMySql("server=127.0.0.1;port=3307;user=root;password=;database=EFCoreIssue22887",
                    b => b.ServerVersion("5.7.0-mysql")
                          .CommandTimeout(18000)) // 5 hours
                .UseLoggerFactory(LoggerFactory.Create(b => b
                    .AddConsole()
                    .AddFilter(level => level >= LogLevel.Information)))
                .EnableSensitiveDataLogging()
                .EnableDetailedErrors();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Module>(
                entity =>
                {
                    entity.Property(e => e.Id)
                        .HasConversion(new GuidToBytesConverter())
                        .IsFixedLength(); // needed, or a varbinary(16) is generated instead of a binary(16)
                    entity.Property(e => e.ModuleTypeId)
                        .HasConversion(new GuidToBytesConverter())
                        .IsFixedLength();
                    
                    // 4. Add in new migration:
                    // entity.Property(e => e.CommonVersionId)
                    //     .HasConversion(new GuidToBytesConverter())
                    //     .IsFixedLength()
                    //     .HasDefaultValue(Guid.Empty);
            
                    entity.HasIndex(e => e.ModuleTypeId);
                });
            
            modelBuilder.Entity<ModuleType>(
                entity =>
                {
                    entity.Property(e => e.Id)
                        .HasConversion(new GuidToBytesConverter())
                        .IsFixedLength();
                    
                    // 4. Add in new migration:
                    // entity.Property(e => e.CommonVersionId)
                    //     .HasConversion(new GuidToBytesConverter())
                    //     .IsFixedLength()
                    //     .HasDefaultValue(Guid.Empty);
                });
        }
    }

    internal static class Program
    {
        private static void Main()
        {
            using var context = new Context();
            
            // 3. Add test data:
            var random = new Random();
            var guidGenerator = new SequentialGuidValueGenerator();
            var moduleTypes = Enumerable.Range(0, 20)
                .Select(n => new ModuleType {Id = guidGenerator.Next(null!)})
                .ToArray();
            var modules =
                Enumerable.Range(0, 1_000_000)
                    .Select(
                        n => DateTime.Today
                            .AddYears(random.Next(-10, 0))
                            .AddDays(random.Next(-365, 0)))
                    .Select(
                        d => new Module
                        {
                            Id = guidGenerator.Next(null),
                            ModuleTypeId = moduleTypes[random.Next(19)].Id,
                            PerformedDate = d,
                            PerformedDistance = random.Next(20_000),
                            Status = random.Next(0, 5),
                            PlannedEndDate = d.AddDays(random.Next(90)),
                            WorkshopScheduleDate = d.AddDays(random.Next(90, 120)),
                        })
                    .ToArray();
            
            context.AddRange(modules);
            context.AddRange(moduleTypes);
            context.SaveChanges(false);
        }
    }
}
```

</details>

I created an initial migration (without the two new columns):

```
PS E:\Sources\PomeloIssues\EFCoreIssue22887\IssueConsoleTemplate> dotnet ef migrations add "Initial"
Build started...
Build succeeded.
warn: Microsoft.EntityFrameworkCore.Model.Validation[10400]
      Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data, this mode should only be enabled during development.
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.8 initialized 'Context' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: CommandTimeout=18000 ServerVersion 5.7.0 MySql SensitiveDataLoggingEnabled DetailedErrorsEnabled
Done. To undo this action, use 'ef migrations remove'
```

Then I updated the database:

```
PS E:\Sources\PomeloIssues\EFCoreIssue22887\IssueConsoleTemplate> dotnet ef database update                                                                                                                                                                                    Build started...
Build succeeded.
warn: Microsoft.EntityFrameworkCore.Model.Validation[10400]
      Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data, this mode should only be enabled during development.
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.8 initialized 'Context' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: CommandTimeout=18000 ServerVersion 5.7.0 MySql SensitiveDataLoggingEnabled DetailedErrorsEnabled
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (3ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      CREATE DATABASE `EFCoreIssue22887`;
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (30ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      CREATE TABLE `__EFMigrationsHistory` (
          `MigrationId` varchar(95) NOT NULL,
          `ProductVersion` varchar(32) NOT NULL,
          CONSTRAINT `PK___EFMigrationsHistory` PRIMARY KEY (`MigrationId`)
      );
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='EFCoreIssue22887' AND TABLE_NAME='__EFMigrationsHistory';
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (2ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      SELECT `MigrationId`, `ProductVersion`
      FROM `__EFMigrationsHistory`
      ORDER BY `MigrationId`;
info: Microsoft.EntityFrameworkCore.Migrations[20402]
      Applying migration '20201006162313_Initial'.
Applying migration '20201006162313_Initial'.
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (29ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      CREATE TABLE `ModuleTypes` (
          `Id` binary(16) NOT NULL,
          CONSTRAINT `PK_ModuleTypes` PRIMARY KEY (`Id`)
      );
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (32ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      CREATE TABLE `Modules` (
          `Id` binary(16) NOT NULL,
          `ModuleTypeId` binary(16) NOT NULL,
          `PerformedDate` datetime(6) NULL,
          `PerformedDistance` int NULL,
          `Status` int NOT NULL,
          `PlannedEndDate` datetime(6) NULL,
          `WorkshopScheduleDate` datetime(6) NULL,
          CONSTRAINT `PK_Modules` PRIMARY KEY (`Id`),
          CONSTRAINT `FK_Modules_ModuleTypes_ModuleTypeId` FOREIGN KEY (`ModuleTypeId`) REFERENCES `ModuleTypes` (`Id`) ON DELETE CASCADE
      );
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (33ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      CREATE INDEX `IX_Modules_ModuleTypeId` ON `Modules` (`ModuleTypeId`);
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (4ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      INSERT INTO `__EFMigrationsHistory` (`MigrationId`, `ProductVersion`)
      VALUES ('20201006162313_Initial', '3.1.8');
Done.
```

I then ran the program to generate 1,000,000 entries for the `Modules` table.

Then I added the two new columns and generated a new migration:

```
PS E:\Sources\PomeloIssues\EFCoreIssue22887\IssueConsoleTemplate> dotnet ef migrations add "AddColumns"                                                                                                                                                                        Build started...
Build succeeded.
Done. To undo this action, use 'ef migrations remove'
```

Finally, I updated the database again, which ran about 5 seconds for the first `ALTER TABLE` statement:

```
PS E:\Sources\PomeloIssues\EFCoreIssue22887\IssueConsoleTemplate> dotnet ef database update                                                                                                                                                                                    Build started...
Build succeeded.
warn: Microsoft.EntityFrameworkCore.Model.Validation[10400]
      Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data, this mode should only be enabled during development.
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.8 initialized 'Context' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: CommandTimeout=18000 ServerVersion 5.7.0 MySql SensitiveDataLoggingEnabled DetailedErrorsEnabled
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (10ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='EFCoreIssue22887' AND TABLE_NAME='__EFMigrationsHistory';
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='EFCoreIssue22887' AND TABLE_NAME='__EFMigrationsHistory';
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (4ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      SELECT `MigrationId`, `ProductVersion`
      FROM `__EFMigrationsHistory`
      ORDER BY `MigrationId`;
info: Microsoft.EntityFrameworkCore.Migrations[20402]
      Applying migration '20201006172246_AddColumns'.
Applying migration '20201006172246_AddColumns'.
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (87ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      ALTER TABLE `ModuleTypes` ADD `CommonVersionId` binary(16) NOT NULL DEFAULT 0x00000000000000000000000000000000;
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (7,039ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      ALTER TABLE `Modules` ADD `CommonVersionId` binary(16) NOT NULL DEFAULT 0x00000000000000000000000000000000;
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (13ms) [Parameters=[], CommandType='Text', CommandTimeout='18000']
      INSERT INTO `__EFMigrationsHistory` (`MigrationId`, `ProductVersion`)
      VALUES ('20201006172246_AddColumns', '3.1.8');
Done.
```

So about 5 seconds for `ALTER TABLE` and 1,000,000 entries. I will try this again with more.

**Comment by incolor2** [comment]
@lauxjpn thanks for relly looking into this. I am running the same version as you Pomelo 3.2.2

I will try running your code against my db. One thing that is diffrent here is that i am running Aurora Serverless

**Comment by incolor2** [comment]
I did take another look at the code now and i realized i have made a mistake

`
            //Register db         
            services.AddDbContext<PlanSearchContext>(options =>
                options.UseMySql(connectionString, mysqlOptions =>
                {
                    //Setting DB version
                    mysqlOptions.ServerVersion(new Version(5, 7), ServerType.MySql);

                    //Some migrations are taking a looong time.
                    mysqlOptions.CommandTimeout(18000);                    
                }

'
In the code we are using the PlanSerarchContext

The PlanSearchContext is also setting the timeout

`        public PlanSearchContext(DbContextOptions<PlanSearchContext> options)
            : base(options)
        {
            Database.SetCommandTimeout(4000); 
        }`
And it seems that is taking precedence. I totally missed that ☹

But still very strange your db is lightning fast and ours is taking hours insted of seconds

**Comment by lauxjpn** [comment]
> The PlanSearchContext is also setting the timeout

@incolor2 Glad you figured out what the issue was! I assumed something similar to that, but its never a bad idea for us to check that the Pomelo implementation does what it is supposed to be doing.

---

I did run the sample code migration again against 10.5 Mio. rows and it took 124 seconds to execute the `ALTER TABLE` statement.

---

Did you try to enable and use the Fast DDL option I linked above? Something like the following should work and might result in better performance (not checked):

1. Add new `Guid?` (nullable) `CommonVersionId` property to your `Module` class (but don't really use them yet in your app) and always initialize the property in your constructor with a `Guid.Empty`. 
2. Add column with fast DDL, which means running `` ALTER TABLE `Modules` ADD `CommonVersionId` binary(16) NULL; `` in your case, after enabling fast DDL (Aurora lab mode). According to the linked Aurora docs, this should execute instantly.
3. Deploy code from step 1.
4. Run `` UPDATE `Modules` SET `CommonVersionId` = 0x00000000000000000000000000000000; ``, because there could have been modules added with `CommonVersionId` set to `NULL` between step 2 and 3.
5. Change `CommonVersionId` property from `Guid?` (nullable) to `Guid` (non-nullable) in your `Module` class (and remove the redundant `Guid.Empty` initialization from the constructor).
6. Run `` ALTER TABLE `Modules` ALTER COLUMN `CommonVersionId` SET DEFAULT 0x00000000000000000000000000000000; ``.
7. Deploy code from step 5.

The SQL parts could be done with migrations, if done in this order.
